### DISCUSSION TITLE
ERROR: failed to ptrace(ATTACH) child 24: Operation not permitted[BUG]

### DISCUSSION THREAD

**Discussion by ixiaofeng** [1Panel-dev/1Panel#5954] [Q&A]
### 联系方式

ixiaofeng.com@gmail.com

### 1Panel 版本

v1.10.13-lts

### 问题描述

间歇性出现 ERROR: failed to ptrace(ATTACH) child 25: Operation not permitted 

### 重现步骤

网上查了一下，说是在docker中部署的php容器默认是不能记录慢日志

### 相关日志输出

```shell
[2K1Panel-php8-UctD  | [26-Jul-2024 16:02:38] ERROR: failed to ptrace(ATTACH) child 25: Operation not permitted (1)
```

**Comment by wanghe-fit2cloud** [comment]
错误 `failed to ptrace(ATTACH) child 25: Operation not permitted` 通常表示 Docker 容器或系统中的进程在尝试使用 `ptrace` 系统调用时遇到权限问题。`ptrace` 是一个用于调试的系统调用，它允许一个进程控制另一个进程的执行。

### 解决方法

1. **检查 Docker 容器的安全设置**

   - 确保 Docker 容器的安全选项允许 `ptrace` 操作。你可以通过在 Docker 启动时添加 `--security-opt` 选项来调整容器的安全设置。例如：
     ```bash
     docker run --security-opt seccomp=unconfined [your-image]
     ```
   - 这将禁用 Seccomp 安全配置文件，允许更广泛的系统调用。请注意，禁用 Seccomp 可能会带来安全风险。

2. **检查系统的安全模块**

   - 某些 Linux 发行版的安全模块（如 SELinux 或 AppArmor）可能会限制 `ptrace` 调用。你可以临时禁用这些安全模块进行测试：
     - **SELinux**:
       ```bash
       sudo setenforce 0
       ```
     - **AppArmor**:
       ```bash
       sudo aa-status
       ```

3. **检查内核参数**

   - 在某些系统上，`ptrace` 操作可能受到内核参数的限制。检查和修改 `/etc/sysctl.conf` 文件中的 `kernel.yama.ptrace_scope` 参数：
     ```bash
     sudo sysctl -w kernel.yama.ptrace_scope=0
     ```

4. **确认容器的权限设置**

   - 确保容器具有足够的权限来执行 `ptrace` 操作。可以使用 `--privileged` 选项来启动容器：
     ```bash
     docker run --privileged [your-image]
     ```
   - 请注意，使用 `--privileged` 选项会赋予容器完全的主机访问权限，因此应谨慎使用。

5. **查看容器日志**

   - 检查容器的详细日志以获取更多信息：
     ```bash
     docker logs <container_id>
     ```

### 示例操作

1. **启动容器时禁用 Seccomp**：
   ```bash
   docker run --security-opt seccomp=unconfined [your-image]
   ```

2. **临时禁用 SELinux**：
   ```bash
   sudo setenforce 0
   ```

3. **修改内核参数**：
   ```bash
   sudo sysctl -w kernel.yama.ptrace_scope=0
   ```

4. **启动容器时使用 `--privileged`**：
   ```bash
   docker run --privileged [your-image]
   ```

这些步骤可以帮助你解决 `ptrace` 相关的权限问题。请根据你的具体情况选择适合的解决方案。
